sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala

```diff
@@ -0,0 +1,258 @@
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, ImplicitCastInputTypes, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.NGrams.NGramBuffer
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, GenericArrayData}
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * Return the top-k n-grams in rows that consist of sequences of strings.
+ */
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr, n, k, accuracy) - Estimates the top-k n-grams in rows that consist of sequences
+      of strings, represented as arrays of strings, or arrays of arrays of strings. 'accuracy' is an
+      optional precision factor that controls memory usage.
+      The parameter 'n' specifies what type of n-grams are being estimated. Unigrams are n = 1, and
+      bigrams are n = 2. Generally, n will not be greater than about 5. The 'k' parameter specifies
+      how many of the highest-frequency n-grams will be returned by the UDAF. The optional precision
+      factor 'accuracy' specifies how much memory to use for estimation; more memory will give
+      more accurate frequency counts, but could crash the JVM. The value will be the max between
+      'accuracy'(0 if it's not specified) and 1000/k, which indicates the max number of n-grams
+      which are kept in the internal HashMap.
+      The output is an array of maps with the top-k n-grams and corresponding frequency.
+  """,
+  extended = """
+    Examples:
+      > SELECT ngrams(array("abc", "abc", "bcd", "abc", "bcd"), 2, 4);
+       [{["abc","bcd"]:2.0},
+       {["abc","abc"]:1.0},
+       {["bcd","abc"]:1.0}]
+  """)
+case class NGrams(
+    child: Expression,
+    nExpression: Expression,
+    kExpression: Expression,
+    accuracyExpression: Expression,
+    override val mutableAggBufferOffset: Int,
+    override val inputAggBufferOffset: Int)
+  extends TypedImperativeAggregate[NGramBuffer] with ImplicitCastInputTypes  {
+
+  def this(child: Expression, nExpression: Expression, kExpression: Expression,
+           accuracyExpression: Expression) = {
+    this(child, nExpression, kExpression, NGrams.getAccuracy(kExpression, accuracyExpression), 0, 0)
+  }
+
+  def this(child: Expression, nExpression: Expression, kExpression: Expression) = {
+    this(child, nExpression, kExpression, Literal(0))
+  }
+
+  private lazy val n: Int = nExpression.eval().asInstanceOf[Int]
+  private lazy val k: Int = kExpression.eval().asInstanceOf[Int]
+  private lazy val accuracy: Int = accuracyExpression.eval().asInstanceOf[Int]
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    Seq(TypeCollection(ArrayType(StringType, false), ArrayType(ArrayType(StringType, false))),
+      IntegerType, IntegerType, IntegerType)
+  }
+
+  val isArrayOfString = child.dataType == ArrayType(StringType, false) ||
+    child.dataType == ArrayType(StringType, true)
```

Comment:
To be honest, I don't there is a way to guarantee no null values at plan compile time.